pkg/network/go/bininspect/dwarf.go

```diff
@@ -0,0 +1,325 @@
+		}
+		locationExpression = loclistEntry.Instr
+	default:
+		return ParameterMetadata{}, fmt.Errorf("unexpected field class on formal parameter's location attribute: locationField=%#v", locationField)
+	}
+
+	totalSize := typ.Size()
+	pieces, err := locexpr.Exec(locationExpression, totalSize, int(d.arch.PointerSize()))
+	if err != nil {
+		return ParameterMetadata{}, fmt.Errorf("error executing location expression for parameter: %w", err)
+	}
+	inspectPieces := make([]ParameterPiece, len(pieces))
+	for i, piece := range pieces {
+		inspectPieces[i] = ParameterPiece{
+			Size:        piece.Size,
+			InReg:       piece.InReg,
+			StackOffset: piece.StackOffset,
+			Register:    piece.Register,
+		}
+	}
+	return ParameterMetadata{
+		TotalSize: totalSize,
+		Kind:      typ.Common().ReflectKind,
+		Pieces:    inspectPieces,
+	}, nil
+}
+
+func (d dwarfInspector) findStructOffsets(structFields []FieldIdentifier) (map[FieldIdentifier]uint64, error) {
+	structOffsets := make(map[FieldIdentifier]uint64)
+	typeReader := dwarfutils.NewTypeFinder(d.dwarfData)
+	for _, fieldID := range structFields {
+		offset, err := typeReader.FindStructFieldOffset(fieldID.StructName, fieldID.FieldName)
+		if err != nil {
+			return nil, fmt.Errorf("could not find offset of \"%s.%s\": %w", fieldID.StructName, fieldID.FieldName, err)
+		}
+		structOffsets[fieldID] = offset
+	}
+	return structOffsets, nil
+}
+
+// getLoclistEntry returns the loclist entry in the loclist
+// starting at offset, for address pc.
+// Adapted from github.com/go-delve/delve/pkg/proc.(*BinaryInfo).loclistEntry
+func (d dwarfInspector) getLoclistEntry(offset int64, pc uint64) (*loclist.Entry, error) {
+	debugInfoBytes, err := godwarf.GetDebugSectionElf(d.elfFile, "info")
+	if err != nil {
+		return nil, err
+	}
+
+	compileUnits, err := dwarfutils.LoadCompileUnits(d.dwarfData, debugInfoBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	debugLocBytes, _ := godwarf.GetDebugSectionElf(d.elfFile, "loc")
+	loclist2 := loclist.NewDwarf2Reader(debugLocBytes, int(d.arch.PointerSize()))
+	debugLoclistBytes, _ := godwarf.GetDebugSectionElf(d.elfFile, "loclists")
+	loclist5 := loclist.NewDwarf5Reader(debugLoclistBytes)
+	debugAddrBytes, _ := godwarf.GetDebugSectionElf(d.elfFile, "addr")
+	debugAddrSection := godwarf.ParseAddr(debugAddrBytes)
+
+	var base uint64
+	compileUnit := compileUnits.FindCompileUnit(pc)
+	if compileUnit != nil {
+		base = compileUnit.LowPC
+	}
+
+	var loclist loclist.Reader = loclist2
+	var debugAddr *godwarf.DebugAddr
+	if compileUnit != nil && compileUnit.Version >= 5 && loclist5 != nil {
+		loclist = loclist5
+		if addrBase, ok := compileUnit.Entry.Val(dwarf.AttrAddrBase).(int64); ok {
+			debugAddr = debugAddrSection.GetSubsection(uint64(addrBase))
+		}
+	}
+
+	if loclist.Empty() {
+		return nil, fmt.Errorf("no loclist found for the given program counter")
+	}
+
+	// Use 0x0 as the static base
+	var staticBase uint64 = 0x0
```

Comment:
Would this only work for statically compiled binaries? Or since we're only ever loading go package symbols, this is not an issue?

Reply:
This entire code is meant to be run on our own test binary, so I think this should not be an issue.